internal/controller/bucket/bucket.go

```diff
@@ -150,105 +144,142 @@ func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.E
 func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
 	bucket, ok := mg.(*v1alpha1.Bucket)
 	if !ok {
 		return managed.ExternalObservation{}, errors.New(errNotBucket)
 	}
 
-	if !c.backendStore.BackendsAreStored() {
-		return managed.ExternalObservation{}, errors.New(errNoS3BackendsStored)
+	if len(bucket.Status.AtProvider.BackendStatuses) == 0 {
+		return managed.ExternalObservation{
+			ResourceExists:   false,
+			ResourceUpToDate: true,
+		}, nil
+	}
+
+	available := false
+	for _, c := range bucket.Status.Conditions {
+		if c.Type == xpv1.TypeReady && c.Reason == xpv1.ReasonAvailable && c.Status == corev1.ConditionTrue {
+			available = true
+
+			break
+		}
+	}
+	if !available && !utils.IsHealthCheckBucket(bucket) {
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: false,
+		}, nil
 	}
 
-	type bucketExistsResult struct {
-		bucketExists bool
-		err          error
+	if len(bucket.Spec.Providers) == 0 {
+		bucket.Spec.Providers = c.backendStore.GetAllActiveBackendNames()
+	}
+
+	allBackendClients := c.backendStore.GetBackendClients(bucket.Spec.Providers)
+
+	missing := len(bucket.Spec.Providers)
+	for _, provider := range bucket.Spec.Providers {
+		if _, ok := allBackendClients[provider]; !ok {
+			missing--
+		}
+		if status := bucket.Status.AtProvider.BackendStatuses[provider]; status == v1alpha1.BackendReadyStatus {
+			missing--
+		}
+	}
+	if missing != 0 {
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: false,
+		}, nil
 	}
 
-	bucketExistsResults := make(chan bucketExistsResult)
+	if !c.backendStore.BackendsAreStored() {
+		return managed.ExternalObservation{}, errors.New(errNoS3BackendsStored)
+	}
 
 	// Create a new context and cancel it when we have either found the bucket
 	// somewhere or cannot find it anywhere.
-	ctxC, cancel := context.WithCancel(ctx)
+	ctxC, cancel := context.WithTimeout(ctx, defaultOperationTimeout)
 	defer cancel()
 
-	if len(bucket.Spec.Providers) == 0 {
-		bucket.Spec.Providers = c.backendStore.GetAllActiveBackendNames()
-	}
+	g := new(errgroup.Group)
 
+	backendError := atomic.Bool{}
 	// Check for the bucket on each backend in a separate go routine
-	allBackendClients := c.backendStore.GetBackendClients(bucket.Spec.Providers)
 	for _, backendClient := range allBackendClients {
-		go func(backendClient *s3.Client, bucketName string) {
-			bucketExists, err := s3internal.BucketExists(ctxC, backendClient, bucketName)
-			bucketExistsResults <- bucketExistsResult{bucketExists, err}
-		}(backendClient, bucket.Name)
+		backendClient := backendClient
+		g.Go(func() error {
+			bucketExists, err := s3internal.BucketExists(ctxC, backendClient, bucket.Name)
+			if err != nil {
+				c.log.Info(errors.Wrap(err, errGetBucket).Error())
+
+				backendError.Store(true)
+
+				// If we have a connectivity issue it doesn't make sense to reconcile the bucket immediately.
```

Comment:
When does the next reconcile get triggered, is it based on a timer or is does it only happen when a change to the CR occurs?

Reply:
It is not implemented in this change. Current version of the feature trusts in backends (once bucket created it is there). But it is pretty easy to write a routine which reconcilies buckets in case of providerconfig health status change. Do you want me to include the (or any other) solution in this PR?

Reply:
WDYT @nolancon ? :point_up:

Reply:
I am fine with that approach @mhmxs but would suggest a separate PR